### PR TITLE
planner: add trace for partition pruning

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number:  ref https://github.com/pingcap/tidb/issues/29661

### What is changed and how it works?

Support trace for partition pruning

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -334,7 +335,7 @@ func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.Part
 		return nil, err
 	}
 	if used != nil {
-		return s.makeUnionAllChildren(ds, pi, convertToRangeOr(used, pi))
+		return s.makeUnionAllChildren(ds, pi, convertToRangeOr(used, pi), opt)
 	}
 	tableDual := LogicalTableDual{RowCount: 0}.Init(ds.SCtx(), ds.blockOffset)
```
Comment (time-and-fate) [inline planner/core/rule_partition_processor.go:340]: Seems we missed this case?

File: planner/core/rule_partition_processor.go [hunk 2]
```diff
@@ -872,24 +873,24 @@ func (s *partitionProcessor) pruneRangePartition(ctx sessionctx.Context, pi *mod
...
+func (s *partitionProcessor) processListPartition(ds *DataSource, pi *model.PartitionInfo, opt *logicalOptimizeOp) (LogicalPlan, error) {
 	used, err := s.pruneListPartition(ds.SCtx(), ds.table, ds.partitionNames, ds.allConds)
 	if err != nil {
 		return nil, err
 	}
 	if used != nil {
-		return s.makeUnionAllChildren(ds, pi, convertToRangeOr(used, pi))
+		return s.makeUnionAllChildren(ds, pi, convertToRangeOr(used, pi), opt)
 	}
 	tableDual := LogicalTableDual{RowCount: 0}.Init(ds.SCtx(), ds.blockOffset)
```
Comment (time-and-fate) [inline planner/core/rule_partition_processor.go:895]: ditto.

**Comment by time-and-fate** [comment]
Also please fix the CI check.

File: planner/core/rule_partition_processor.go [hunk 3]
```diff
@@ -1568,3 +1577,53 @@ func (p *rangeColumnsPruner) pruneUseBinarySearch(sctx sessionctx.Context, op st
...
+	var used []model.PartitionDefinition
+	for _, def := range usedMap {
+		used = append(used, def)
+	}
+	sort.Slice(used, func(i, j int) bool {
+		return used[i].ID < used[j].ID
+	})
+	if len(children) == 1 {
+		action = fmt.Sprintf("Datasource[%v] becomes %s[%v]", ds.ID(), plan.TP(), plan.ID())
+		reason = fmt.Sprintf("Datasource[%v] has one available partiton table[%s] after partition pruning", ds.ID(), used[0].Name)
```
Comment (time-and-fate) [inline planner/core/rule_partition_processor.go:1597]:
```suggestion
		reason = fmt.Sprintf("Datasource[%v] has one needed partiton[%s] after partition pruning", ds.ID(), used[0].Name)
```

File: planner/core/rule_partition_processor.go [hunk 4]
```diff
@@ -1568,3 +1577,53 @@ func (p *rangeColumnsPruner) pruneUseBinarySearch(sctx sessionctx.Context, op st
...
+			buffer.WriteString("] after partition pruning")
+			return buffer.String()
+		}()
+	}
+	opt.appendStepToCurrent(ds.ID(), ds.TP(), reason, action)
+}
+
+func appendNoPartitionChildTraceStep(ds *DataSource, dual LogicalPlan, opt *logicalOptimizeOp) {
+	action := fmt.Sprintf("Datasource[%v] becomes %v[%v]", ds.ID(), dual.TP(), dual.ID())
+	reason := fmt.Sprintf("Datasource[%v] has no available partition table after partition pruning", ds.ID())
```
Comment (time-and-fate) [inline planner/core/rule_partition_processor.go:1627]:
```suggestion
	reason := fmt.Sprintf("Datasource[%v] doesn't have needed partition after partition pruning", ds.ID())
```

File: planner/core/rule_partition_processor.go [hunk 5]
```diff
@@ -1568,3 +1577,53 @@ func (p *rangeColumnsPruner) pruneUseBinarySearch(sctx sessionctx.Context, op st
...
+				if i > 0 {
+					buffer.WriteString(",")
+				}
+				buffer.WriteString(fmt.Sprintf("%s[%v]", child.TP(), child.ID()))
+			}
+			buffer.WriteString("]")
+			return buffer.String()
+		}()
+		reason = func() string {
+			buffer := bytes.NewBufferString(fmt.Sprintf("Datasource[%v] have multi available partition tables[", ds.ID()))
```
Comment (time-and-fate) [inline planner/core/rule_partition_processor.go:1611]:
```suggestion
			buffer := bytes.NewBufferString(fmt.Sprintf("Datasource[%v] has multiple needed partitions[", ds.ID()))
```

**Review by time-and-fate — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]

**Pull request merged**